src/ai/defaultai.cc

```diff
@@ -305,43 +303,30 @@ void DefaultAI::think() {
 		scheduler_delay_counter_ = 0;
 	}
 
-	if (jobs_to_run_count == 0) {
-		// well we have nothing to do now
-		return;
-	}
-
 	// And printing it now and resetting counter
 	if (scheduler_delay_counter_ > 10) {
 		log_warn_time(gametime, " %d: AI: game speed too high, jobs are too late (now %2d seconds)\n",
 		              player_number(), static_cast<int32_t>(delay_time / 1000));
 		scheduler_delay_counter_ = 0;
 	}
 
-	// 400 provides that second job is run if delay time is longer then 1.6 sec
```

Comment:
these lines and especially the comment are not up to date.   
currently second job is run if delay > 2000 and third one if delay > 4500 and fourth if delay > 8000
so this if could safely evaluate for delay >=2000 to spare some sqrt calcualtions.     
furthermore these values would make for good values in our auto_speed function where we decide the max delay allowed and by this the max jobs to be run. so I believe 5000 wasn't that bad a value as it allows for the third job to be run.

Reply:
Yes, but I propose simpler logic - first of all, all overdue jobs deserve to be run, however we have to limit how many of them will be run - for performance reasons. 
I am not sure your understanding of the old code is right. It is my code BTW, so I remember my intend: 'the more delayed are the more jobs to be executed.'

Reply:
Fully understood this. And when running the game at max. speed there might be not much difference but in a normal case of a game run at 2x or 3x max there might be no delays at all and running 3 jobs at once would cause a visible glitch I'm afraid.
So your new design would be fine as long as the jobs are increased starting with one. Function maybe linear instead of quadratic though to simplify calculations .

Reply:
So, if we reduce these 500 ms to 200 and jobs that might be run in one think() to 1 or 2, this is what you would prefer?

Reply:
Not really.   
as long as the ai is in the same thread with the logic code it has to share computing time with logic. The UI is drawn now in its own thread so the fps stay up which is good. But this makes it clearly visible when lenghty prozesses run preventing the logic of the game being updated (this was the case letting frames drop previously). 
So running more jobs in the same amount of time would help AI Training probably but would have effects worsening #4963.
So I would keep the 500ms. and start the function to determine jobs to run with 1 job. Only if delay builds up (due to increasing game speed or lenghty calcualtions in the logic due to game size) we should increase the jobs to run. this can be done in a linear way. 
So to sum it up we could use your new layout but you might consider implementing my suggestion.

Reply:
After sorting of tasks, delays can be:
5500, 4800, 4200, 4100, 2800, 2200....
vs 
5500, 1200, 300, -500, -800, -2100
So the delay of first member is not ideal quantification how delayed we are.

Let say we check 5 military sites (no matter of total count) for attack every 10 seconds and here we say, sorry - every 20 seconds it will be, while enemy military sites count grows, it will detoriate performance of AI, though you will have preserves fluentness of the game.

Reply:
There is some internal inconsistency - one piece of AI code says - this task should be executed each 10 seconds, and other piece of code will say - it can be delayed by 5 seconds and nobody cares...

Reply:
What about to have minimal delay for jobs execution
for first job if delayed
for second if delayed for more than 300 ms
for third if delayed for more than 600 ms
and so on

Reply:
I agree on the first delay not ideal to measure delay. However it is a very good indication and if we are really delayed (your example 1) we would be delayed in the next cycle as well causing multiple jobs to run again. 
As said we check every 500 ms so the last task not processed in the list will be delayed 500 ms realtime which will be 5 secs at 10x gamespeed. 
I think we aim at an basic interval for each task but we can delay each of them by a certain amount of time. This is why the algorithm in question takes place. we start to do something against delay only if we are 2 seconds gametime late.

Reply:
500s is gametime, or not? If realtime than this is very bad..

Reply:
So would you agree with:

1. we define jobs to run (can be hardcoded 5)
2. we will iterate over task queue (up to defined count above) and job will be executed if 
delay > position * 300
where position starts with 0

Reply:
it is realtime (and must be) as if it would be game time we would need to run an AI think cycle every 50 ms at 10x gamespeed where some of our tasks last for several 10ms (especially update all buildable fields is costly)

Reply:
Some information that some tasks collects has "expiration time" and it is expected that another tasks that uses this information are run while the information are not expired yet. This is probably why AI will not build anything in high speed - tasks deciding about buildings does not have "fresh" information about buildable spots. - Just general explanation

Reply:
BTW, AI newer update ALL buildable fields in one run of particular task. There is limit, ca 50 per one run of the task. But negative thing is that if you have 2000 buildable tasks ... we do your math :)

Reply:
I never did any time profiling of the tasks, would be interesting to know and have real info, but I really tried to keep them from going 'wild'

Reply:
I know about the fields ( i recently tried to make this thread safe in my attempt to run the AI in its own thread). the problem why this is costly is we itrate over about 50 fields but for each field we check the fields in a radius of 8 for economy up to 22 for military aspects and we are doing this for each player. which is a lot of iterations obviously.  
in our current design (having the same thread as the logic code) the interval of the 2 tasks should ensure for the info not expire however if the gamespeed will set to progress to fast we will not be able to calculate everything needed, but the logic won't be able to do so neither so we are stalling the game.

Reply:
We can check less sites, more frequently. Now the frequency is:
constexpr Duration kMinBFCheckInterval(5 * 1000);
Of course 5 seconds are never obeyed with this our scheduler

Reply:
Generally if we do not measure something, than this is just a quesswork. Maybe there are bigger resource hogs in task than this update buildable fields. Maybe some cumulative CPU counter for each task would be nice, to get stats for a hour or so...

Reply:
in my attempts to make AI threading working I tried to set Mutex locks blocking the logic code in this function and it blocked the logic for a time clearly visible in the game. However logging SDL_ticks() at htne beginning an the end of a task might be a good idea to measure our code and identify performance improvements.

Reply:
sounds good. maybe SDL_ticks is not precise enough.